File: child.js
#!/usr/bin/env node

'use strict';

var Module = require('module'),
    path = require('path'),
    fs = require('fs'),
    net = require('net');

// connect to the parent process at the socket address passed in

const socketPath = process.argv[2];

const socket = new net.Socket();

socket.connect(socketPath, () => {

    // tap into Module._load and various fs methods to
    // notify the parent whenever a file is required,
    // imported, or read in to the process

    var load = Module._load;
    Module._load = function (p, mod) {
        var fullPath = Module._resolveFilename(p, mod, false);
        if (fullPath)
            socket.write(fullPath + '\n', 'utf8');

        return load.apply(this, arguments);
    };

    // The following will be used to watch other static
    // files used by the process - like .html files.
    // It should only send back paths the process does
    // not change itself - temp files should not be watched.
    // TODO: this is quite complex to work out, so for
    // now provide 'and watch this directory' functionality

    // [
    //     'readFileSync',
    //     'readFile',
    //     'statSync',
    //     'stat'
    // ].forEach(function (method) {
    //     var m = fs[method];
    //     fs[method] = function (p) {
    //         process.send(path.resolve(p));
    //         return m.apply(this, arguments);
    //     };
    // });

    // remove this file and the socket address from the arguments
    // so it appears that the js file has been called directly

    process.argv.splice(1, 2);

    // require the first argument directly to run it with
    // modified require() and import

    require(path.resolve(process.argv[1]));
});

File: child.py
#!/usr/bin/env python

import sys
import socket
import __builtin__
import re

# connect to the parent process using the socket address passed in

socket_file = sys.argv[1]

s = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)

s.connect(socket_file)

# cache files that are already required in a dictionary

sent = {}

# override __import__, sending the paths of
# newly required modules to the parent

_import = __builtin__.__import__

def importOverride(*args, **kwargs):
    mod = _import(*args, **kwargs)
    if hasattr(mod, '__file__') and mod.__file__ not in sent:
        # mark as processed
        sent[mod.__file__] = True
        s.send(re.sub(r'\.pyc$', '.py', mod.__file__) + '\n');
    return mod

__builtin__.__import__ = importOverride

# remove this file and the socket address from the arguments,
# and execute the underlying file

sys.argv = sys.argv[2:]
execfile(sys.argv[0])
